core/query/sql.ml

```diff
@@ -137,138 +152,162 @@ let order_by_clause n =
 
       (* optimisation *)
-      | Apply ("not", [Empty q]) -> "exists (" ^ string_of_query quote true q ^ ")"
+      | Apply (op, [v; w]) when StringSet.mem op binary_ops -> Format.fprintf ppf "(%a) %s (%a)" pr_b v (binary_map op) pr_b w
+      | Apply (f, args) when SqlFuns.is f ->
+        let pp_value ppf x = Format.fprintf ppf "%a" pr_b x in
+        Format.fprintf ppf "%a(%a)" pp_constant (SqlFuns.name f) (Format.pp_print_list ~pp_sep:pp_comma pp_value) args
+      | Apply (f, args) ->
+        let pp_value ppf x = Format.fprintf ppf "%a" pr_b x in
+        Format.fprintf ppf "%a(%a)" pp_constant f (Format.pp_print_list ~pp_sep:pp_comma pp_value) args
+      | Empty q -> Format.fprintf ppf "not exists (%a)" pr_q q
+      | Length q -> Format.fprintf ppf "select count(*) from (%a) as %a" pr_q q pp_constant (fresh_dummy_var ())
+      | RowNumber [] -> Format.fprintf ppf "%a" pp_constant "1"
+      | RowNumber ps -> Format.fprintf ppf "row_number() over (order by %a)" pp_constant (String.concat "," (List.map (string_of_projection quote one_table) ps))
 
-      | Apply ("not", [v]) -> "not (" ^ sb v ^ ")"
-      | Apply (("negate" | "negatef"), [v]) -> "-(" ^ sb v ^ ")"
-      | Apply ("&&", [v; w]) -> "(" ^ sb v ^ ")" ^ " and " ^ "(" ^ sb w ^ ")"
-      | Apply ("||", [v; w]) -> "(" ^ sb v ^ ")" ^ " or " ^ "(" ^ sb w ^ ")"
-      | Apply ("==", [v; w]) -> "(" ^ sb v ^ ")" ^ " = " ^ "(" ^ sb w ^ ")"
-      | Apply ("<>", [v; w]) -> "(" ^ sb v ^ ")" ^ " <> " ^ "(" ^ sb w ^ ")"
-      | Apply ("<", [v; w]) -> "(" ^ sb v ^ ")" ^ " < " ^ "(" ^ sb w ^ ")"
-      | Apply (">", [v; w]) -> "(" ^ sb v ^ ")" ^ " > " ^ "(" ^ sb w ^ ")"
-      | Apply ("<=", [v; w]) -> "(" ^ sb v ^ ")" ^ " <= " ^ "(" ^ sb w ^ ")"
-      | Apply (">=", [v; w]) -> "(" ^ sb v ^ ")" ^ " >= " ^ "(" ^ sb w ^ ")"
-      | Apply ("RLIKE", [v; w]) -> "(" ^ sb v ^ ")" ^ " RLIKE " ^ "(" ^ sb w ^ ")"
-      | Apply ("LIKE", [v; w]) -> "(" ^ sb v ^ ")" ^ " LIKE " ^ "(" ^ sb w ^ ")"
-      | Apply (f, args) when SqlFuns.is f -> SqlFuns.name f ^ "(" ^ String.concat "," (List.map sb args) ^ ")"
-      | Apply (f, args) -> f ^ "(" ^ String.concat "," (List.map sb args) ^ ")"
-      | Empty q -> "not exists (" ^ string_of_query quote true q ^ ")"
-      | Length q -> "select count(*) from (" ^ string_of_query quote true q ^ ") as " ^ fresh_dummy_var ()
-      | RowNumber [] -> "1"
-      | RowNumber ps ->
-        "row_number() over (order by " ^ String.concat "," (List.map (string_of_projection quote one_table) ps) ^ ")"
-and string_of_projection quote one_table (var, label) =
-  if one_table then
-    quote label
-  else
-    string_of_table_var var ^ "." ^ (quote label)
+let string_of_base quote one_table b =
+  let pr_b ppf b = pr_base ppf quote one_table b in
+  Format.asprintf "%a" pr_b b
 
 let string_of_query ?(range=None) quote q =
-  let range =
+  let pr_q ppf q = pr_query ppf quote false q in
```

Comment:
This definition of pr_q is used exactly once.

It might be an idea to rearrange the arguments to pr_query to be pr_query quote b ppf q, so that instead of using this custom definition of pp_q (or having to write something like (fun ppf -> pr_query ppf quote false q) you can write

```
Format.asprintf "%a%a" (pr_query quote false) q
```
that is, so that the last two arguments are ppf and q and the behavior-modifying  parameters come before ppf.